과제 3 MVP

namget opened the pull request:
과제 3 MVP 
MVP 내용이 조금 난해하네요 어디까지 View로 처리해야할지..
일단 제출해봅니다.

Commit:
- Presenter 생성중

Commit:
- 각 Presenter 생성
- 각 Contract 생성
- BasePresenter 생성
- makeToast -> ShowToast로 이름 변경
- main화면 ProgressBar 추가

Commit:
- present 안쓰는부분 선언 에러 수정

preludezdev commented on namget/app/src/main/java/com/namget/myarchitecture/ui/repo/RepoContract.kt:
```diff
@@ -0,0 +1,21 @@
+interface RepoContract{
+    interface View : BaseView<Presenter> {
+        fun setUserInfoData(userInfoResponse: UserInfoResponse)
+        fun setRepoInfoData(repoInfoResponse: RepoInfoResponse)
```
MVP 의 View 에서 함수 수행 결과가 데이터 set에서 딱 끝나면 네이밍 prefix 에 set 을 붙이는데 수행결과가 ui 에 보여주는거까지 이어지는 작업이면 보통 show 를 붙이더라구요. 이전 스터디때 저도 set 으로 했다가 이런 함수들은 이름 prefix를 show 로 바꿔야 한번에 와닿는다고 해서 show를 쓰고 있습니다. 구글 블루프린트에도 보면 show로 시작하는 함수가 많으니 한번 참고만 해보세요. 일종의 컨벤션 같은건가봐요.

namget replied:
수정했습니다.

preludezdev commented on namget/app/src/main/java/com/namget/myarchitecture/ui/repo/RepoActivity.kt:
```diff
@@ -47,40 +84,14 @@ class RepoActivity : BaseActivity(), BaseView {
+        return super.onOptionsItemSelected(item)
     }
 
     companion object {
```
안쓰는 태그는 지워주세요

namget replied:
수정했습니다.

preludezdev commented on namget/app/src/main/java/com/namget/myarchitecture/ui/repo/RepoActivity.kt:
```diff
@@ -47,40 +84,14 @@ class RepoActivity : BaseActivity(), BaseView {
         }
+        return super.onOptionsItemSelected(item)
     }
 
```
RepoActivity 에도 액티비티 라이프사이클에 맞춰 Disposable 객체 dispose 해주세요

namget replied:
수정했습니다.

preludezdev replied:
BaseActivity 에서 해당 부분을 처리하셨던데 로직이 간단한 액티비티의 경우 (ex. 스플래쉬) 는 Presenter 가 없을 수도 있지 않을까요?

namget replied:
그정도로 비즈니스 로직이 필요하지 않는 경우에는 베이스액티비티를 따로 상속을 안받고 처리해도 되지 않을까요

preludezdev replied:
그것도 맞는 말이네요. 그럼 이 부분은 스타일 차이로 이해하겠습니다.

preludezdev commented on namget/app/src/main/java/com/namget/myarchitecture/ui/main/MainActivity.kt:
```diff
@@ -47,25 +56,20 @@ class MainActivity : BaseActivity() {
-
     companion object {
         private const val TAG = "MainActivity"
     }
```
안쓰는 태그는 지워주세요 222

namget replied:
수정했습니다.

preludezdev commented on namget/app/src/main/java/com/namget/myarchitecture/ui/main/MainActivity.kt:
```diff
@@ -47,25 +56,20 @@ class MainActivity : BaseActivity() {
             startActivity(Intent(this, SearchActivity::class.java))
         }
     }
 
```
메인액티비티도 dispose 처리해주세요 222

namget replied:
수정했습니다.

preludezdev replied:
이하동문

preludezdev commented on namget/app/src/main/java/com/namget/myarchitecture/ui/base/BaseActivity.kt:
```diff
@@ -18,5 +18,4 @@ abstract class BaseActivity : AppCompatActivity() {
     }
 
-
 }
```
- BaseActivity 에 있는 disposable 객체는 쓰이지 않고 있는 것 같습니다. 아래 onDestroy() 랑 같이 정리해주세요
- 지금은 모든 액티비티가 repository 를 필요로 하고 있지만 새로 생기는 액티비티 클래스도 repository 를 필요로 하는지 확신하시나요?

namget replied:
- 수정했습니다.

preludezdev commented on namget/app/src/main/java/com/namget/myarchitecture/ui/search/SearchActivity.kt:
```diff
@@ -54,10 +56,8 @@ class SearchActivity : BaseActivity() , BaseView {
+            searchPresenter.insertRepoData(searchAdapter.getAdapterItem(it))
             //detail
             startRepoActivity(
                 searchAdapter.getAdapterItem(it).fullName,
```
getAdapterItem(it) 이 세번 호출되고 있습니다. 변수로 저장해서 한번만 찾게 하는건 어떨까요?

namget replied:
수정했습니다

preludezdev commented on namget/app/src/main/java/com/namget/myarchitecture/ui/search/SearchActivity.kt:
```diff
@@ -118,7 +124,7 @@ class SearchActivity : BaseActivity() , BaseView {
                 if (!query.isNullOrEmpty()) {
                     hideKeyboard()
                     showDialog()
```
개인적으로 View 인터페이스에 정의하는 함수들은 프레젠터에서 쓰기위해 정의한다고 생각하는데 이렇게 뷰에서 직접 사용하게 되면 Contract 인터페이스를 정의하는 의미가 사라지는 것 같습니다

namget replied:
이렇다는 말씀은 해당부분을 프레젠터에도 똑같이 만들고 다시 프레젠터에서 뷰로 호출하는 형태로 진행해야한다는 말씀이신가요?
지금 text submit이라는 callback은 액티비티에 종속되어있을수도 있는데 프래젠터에서 처리하기 위해서는 프레젠터의 생성자 혹은 setter를 통해 callback 추상화 리스너를 넘겨야하는데 
그럼 어떻게 처리하는게 좋을까요?

preludezdev replied:
제가 말하고 싶은 부분은 익명 클래스 부분이 아니고 지금 익명 클래스 안에 있는 hideKeyboard() 함수와 showDialog() 함수가 hideDialog() 함수가 쓰인 것처럼 Presenter 의 requestRepoList() 함수 안에 있어야되지 않을까? 였습니다.

namget replied:
수정했습니다 f8e6c095572193d01868a392110c0e90c0a8d17a

preludezdev reviewed (commented):
안녕하세요 남겟님,
부족하지만 피드백 남겼습니다. 잘 부탁드립니다 :)

Commit:
- set -> show로 컨벤션 맞춤
- 안쓰는 TAG 삭제
- dispose 라이프사이클 Base가 관리하도록 변경
- getAdapter로 중복 불리는부분 수정
- BaseActivty disposable 안쓰이는 부분 삭제

Commit:
- core package 삭제

Commit:
- RepoRepositoryInf를 통해 RepoRepository가 필요한 클래스에 필수구현

preludezdev commented:
수정하고 나서 수정했다고 말할 때 수정한 커밋을 함께 달아주면 피드백이 어떻게 반영됐는지 확인하기 더 좋을 것 같습니다 :)

Commit:
- request함수 안에 뷰의 키보드와 다이어로그 hide 기능 이동

preludezdev reviewed (approved):
LGTM!!
고생하셨습니다 :)
